### PR TITLE
Troubleshooting sync/fingerprint/scan

### PR DESCRIPTION
**Pull request by gvwilson**
Troubleshooting problems with database sync and scan.

### CODE REVIEW

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+3. Can you connect to the data warehouse from another client from the machine you're running Metabase on?
+4. Can you run a query like the one shown below (which returns an empty set of results) for each table you're supposed to be able to access?
+
+```
+SELECT TRUE
+FROM table
+WHERE 1 <> 1
+LIMIT 0
+```
+
```
Comment (flamber) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:32]:
```suggestion
Please note that we only get the first 10,000 documents in a MongoDB collection, so if you are not seeing new fields, then the fields might not exist in those documents. Please see [this discussion][metabase-mongo-missing] for more details on MongoDB in particular.

```

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+Many relational databases allow administrators to add descriptions to columns when creating tables. These descriptions are separate from Metabase's internal metadata, but when Metabase syncs with the database, it loads these descriptions so that they can be displayed along with our metadata.
+
+**How to detect this:** Descriptions that are in the database aren't showing up.
+
+**How to fix this:**
+
+1. The usual cause is that the descriptions don't exist: most database administrators don't add descriptions to columns.
+2. Additionally, Oracle manages descriptions differently from other relational databases, and non-relational databases such as MongoDB don't have column descriptions at all.
+3. If you are sure that the descriptions are there and you are using a relational database other than Oracle, you may have found a bug: please [let us know][bugs].
+
```
Comment (flamber) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:63]: I would remove this. I've only seen this being a problem once - on Oracle.
```suggestion
```

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+3. Check the error message returned from Metabase.
+4. Check the credentials you are using to authenticate and make sure they identify your script as a user with administrative privileges.
+
+### Sync and scan take a very long time to run
+
+**How to detect this:** Sync and scan take longer than a few seconds to complete, and/or the UI freezes while you are waiting for them to finish.
+
+**How to fix this:** The root cause is usually an unusual database schema. For example, if you have a table with several thousand columns (not rows---columns), it will take Metabase a while to scan them all. While this may seem unlikely, it can sometimes happen when a table has been pivoted so that (for example) customer IDs have been turned into columns.
+
+You can "fix" this by disabling scan entirely by going to the database in the Admin Panel and telling Metabase, "This is a large database," and then going to the Scheduling tab. However, sync is necessary: without it, Metabase won't know what tables exist or what columns they contain. To actually solve the problem, you will have to re-think the organization of your database.
```
Comment (flamber) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:87]: Those processes are async, so they cannot freeze the UI (anymore).
```suggestion
**How to detect this:** Sync and scan take a long time to complete.

**How to fix this:** 
1. For sync, this is usually caused by a large database with hundreds of schema, thousands of table and with hundreds of columns in each table. If only few of those are needed in Metabase, then restricting the privileges used to connect to the database with will make sure that Metabase can only sync a limited subset of the database.
2. Scanning normally takes longer than sync, but changing the amount of fields which has "Filtering on this field" set to "A list of all values" to either "Search box" or "Plain input box" will mean that those will not be included in the scan.

You can "fix" this by disabling scan entirely by going to the database in the Admin Panel and telling Metabase, "This is a large database," and then going to the Scheduling tab. However, sync is necessary: without it, Metabase won't know what tables exist or what columns they contain.
```

File: docs/troubleshooting-guide/index.md [hunk 4]
```diff
@@ -32,7 +32,9 @@ Problems, their causes, how to detect them, and how to fix them.
 
 - [Handling timezones][timezones].
 
-- [Manage data sandboxing][sandbox]
+- [Managing data sandboxing][sandbox].
+
+- [Synchronizing with the database][sync-fingerprint-scan]
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/index.md:37]: Yeah - @jeff-bruemmer and I are trying to come up with expressive titles and subtitles. I'll add this one...

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# Synchronizing with the database
+
+Metabase needs to know what's in your database in order to populate dropdown menus and suggest good visualizations, but loading all the data would be very slow (or simply impossible if you have a lot of data). It therefore does three things:
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:3]: Thanks.

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+1. Metabase periodically asks the database what tables are available, then asks which columns are available for each table. We call this *syncing*, and happens every hour unless you configure it to run more or less frequently. It's very fast with most relational databases, but can be slower with MongoDB, Athena, and Presto back-ends.
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:5]: Reworded.

**Comment by paoliniluis** [comment]
I'm OK with this one, we have a few here and theres with Mongo that we know of (like Atlas that does not want to show the tables), but it's a matter of time till we iron those out

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+2. Check the URL you are using to send the request to Metabase.
+3. Check the error message returned from Metabase.
+4. Check the credentials you are using to authenticate and make sure they identify your script as a user with administrative privileges.
+
+### Sync and scan take a very long time to run
+
+**How to detect this:** Sync and scan take a long time to complete.
+
+**How to fix this:** 
+1. For sync, this is usually caused by a large database with hundreds of schema, thousands of table and with hundreds of columns in each table. If only few of those are needed in Metabase, then restricting the privileges used to connect to the database with will make sure that Metabase can only sync a limited subset of the database.
```
Comment (jeff-bruemmer) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:71]:
```suggestion
1. For sync, delays are usually caused by a large database with hundreds of schema, thousands of table and with hundreds of columns in each table. If you only need a subset of those tables or columns in Metabase, then restricting the privileges used to connect to the database will make sure that Metabase can only sync a limited subset of the database.
```

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+2. Scanning normally takes longer than sync, but changing the amount of fields which has "Filtering on this field" set to "A list of all values" to either "Search box" or "Plain input box" will mean that those will not be included in the scan.
```
Comment (jeff-bruemmer) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:72]:
```suggestion
2. Scanning normally takes longer than sync, but you can reduce the number of fields Metabase will scan by changing the number of fields that have the **Filtering on this field** option set to "A list of all values". Setting fields to either "Search box" or "Plain input box" will exclude those fields from scans.
```

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+You can "fix" this by disabling scan entirely by going to the database in the Admin Panel and telling Metabase, "This is a large database," and then going to the Scheduling tab. However, sync is necessary: without it, Metabase won't know what tables exist or what columns they contain.
```
Comment (jeff-bruemmer) [inline docs/troubleshooting-guide/sync-fingerprint-scan.md:74]: Do we even want to recommend disabling scans?

**Review by jeff-bruemmer — APPROVED** [review]
Nice, succinct guide, @gvwilson. Just some small notes. Main note is we should probably frontload the "Metabase isn't showing all of the values I expect to see" problem, as that's likely the trouble they're here to shoot.

File: docs/troubleshooting-guide/sync-fingerprint-scan.md (hunk 9 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:74]: Johannes wanted it in with the disclaimer.

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+**How to fix this:**
+
+1. Go to the Admin Panel and select "Data Model"
+2. Select the database, schema, table and field in question.
+3. Click the gear-icon to view all the field's settings.
+4. Set "Field Type" to "Category" and "Filtering on this field" to "A list of all values"
+5. Click the button "Re-scan this field" in the bottom.
+
+### I cannot force Metabase to sync or scan using the API
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:49]: yes - I'll add

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+2. Metabase *fingerprints* the column the first time it synchronizes. Fingerprinting fetches the first 10,000 rows from each column and uses that data to guesstimate how many unique values each column has, what the minimum and maximum values are for numeric and timestamp columns, and so on. Metabase only fingerprints each column once unless the administrator explicitly tells it to run again, or when there has been changes to the fingerprinting logic in newer versions (happens rarely).
+
+3. *Scan* is similar to fingerprinting, but is done every 24 hours (unless it's configured to run less often or disabled).  Scanning looks at the first 5000 distinct records ordered ascending, when a field is set to "A list of all values" in the Data Model, which is used to display options in dropdowns. If the textual result of scanning a column is more than 10 kbyte long, for example, we display a search box instead of a dropdown.
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:9]: included

File: docs/troubleshooting-guide/sync-fingerprint-scan.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+If the credentials Metabase is using to connect to the database don't give it privileges to read the tables, the first sign will often be a failure to sync, which would then also stop fingerprint and scan.
+
+**How to detect this:** You are unable to see any of the tables in the database, or column that have just been added don't show up in Metabase.
+
+**How to fix this:** [This guide][troubleshooting-db-connection] explains how to troubleshoot database connections. The relevant steps for solving this problem are:
+
+1. If you have just set up a new database in Metabase, the sync process might still be running---it's normally fast, but does sometimes take a while. You can follow its progress in Admin > Troubleshooting > Logs.
+2. If you have just added a table or a column, Metabase might not have synced yet. You can manually run the sync process by going to the Admin Panel, selecting "Databases", choosing your database, and clicking on "Sync database schema now".
+3. Sometimes browsers will show an old cached list of tables or columns. Refreshing the page will update the cache.
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/sync-fingerprint-scan.md:23]: yup

**Pull request merged**